src/parser/Literal.cpp

```diff
@@ -48,31 +48,44 @@ NormalizedStringView Literal::getLanguageTag() const {
 }
 
 // __________________________________________
-Literal Literal::literalWithQuotes(const std::string& contentWithQuotes) {
-  return Literal(RdfEscaping::normalizeLiteralWithQuotes(contentWithQuotes));
-}
-Literal Literal::literalWithoutQuotes(const std::string& contentWithoutQuotes) {
-  return Literal(
-      RdfEscaping::normalizeLiteralWithoutQuotes(contentWithoutQuotes));
-}
-Literal Literal::literalWithQuotesWithDatatype(
-    const std::string& contentWithQuotes, Iri datatype) {
-  return {RdfEscaping::normalizeLiteralWithQuotes(contentWithQuotes),
-          std::move(datatype)};
-}
-Literal Literal::literalWithoutQuotesWithDatatype(
-    const std::string& contentWithoutQuotes, Iri datatype) {
-  return {RdfEscaping::normalizeLiteralWithoutQuotes(contentWithoutQuotes),
-          std::move(datatype)};
+Literal Literal::literalWithQuotes(
+    const string& rdfContentWithQuotes,
+    const std::optional<std::variant<Iri, string>>& descriptor) {
+  NormalizedString content =
+      RdfEscaping::normalizeLiteralWithQuotes(rdfContentWithQuotes);
+
+  return literalWithNormalizedContent(content, descriptor);
 }
-Literal Literal::literalWithQuotesWithLanguageTag(
-    const std::string& contentWithQuotes, const std::string& languageTag) {
-  return {RdfEscaping::normalizeLiteralWithQuotes(contentWithQuotes),
-          RdfEscaping::normalizeLanguageTag(languageTag)};
+
+// __________________________________________
+Literal Literal::literalWithoutQuotes(
+    const string& rdfContentWithoutQuotes,
+    const std::optional<std::variant<Iri, string>>& descriptor) {
+  NormalizedString content =
+      RdfEscaping::normalizeLiteralWithoutQuotes(rdfContentWithoutQuotes);
+
+  return literalWithNormalizedContent(content, descriptor);
 }
-Literal Literal::literalWithoutQuotesWithLanguageTag(
-    const std::string& contentWithoutQuotes, const std::string& languageTag) {
-  return {RdfEscaping::normalizeLiteralWithoutQuotes(contentWithoutQuotes),
-          RdfEscaping::normalizeLanguageTag(languageTag)};
+
+// __________________________________________
+Literal Literal::literalWithNormalizedContent(
+    const NormalizedString& normalizedRdfContent,
+    const std::optional<std::variant<Iri, string>>& descriptor) {
+  if (!descriptor.has_value()) {
+    return Literal(normalizedRdfContent);
+  }
+
+  if (std::holds_alternative<std::string>(descriptor.value())) {
+    NormalizedString languageTag =
+        RdfEscaping::normalizeLanguageTag(std::get<string>(descriptor.value()));
+    return {normalizedRdfContent, languageTag};
+  }
+
+  else if (std::holds_alternative<Iri>(descriptor.value())) {
+    return {normalizedRdfContent, std::get<Iri>(descriptor.value())};
+  }
+
+  AD_THROW("Descriptor variable holds unsupported value type.");
```

Comment:
Use `std::visit()` instead of the manual `if/else if` that way you get the static guarantees that you miss no case and don't need the runtime AD_THROW`.

Reply:
(either with `if constexpr ... else` (see cppreference) or using `ad_utility::OverloadCallOperator` (many usages inside the project).